io/src/read_detector_description.cpp

```diff
@@ -0,0 +1,175 @@
+/** TRACCC library, part of the ACTS project (R&D line)
+ *
+ * (c) 2024 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+// Library include(s).
+#include "traccc/io/read_detector_description.hpp"
+
+#include "csv/read_surfaces.hpp"
+#include "traccc/io/read_detector.hpp"
+#include "traccc/io/read_digitization_config.hpp"
+#include "traccc/io/utils.hpp"
+
+// VecMem include(s).
+#include <vecmem/memory/host_memory_resource.hpp>
+
+// System include(s).
+#include <sstream>
+#include <stdexcept>
+
+namespace {
+
+void fill_digi_info(traccc::silicon_detector_description::host& dd,
+                    const traccc::module_digitization_config& data) {
+
+    // Set a hard-coded threshold for which cells should be considered for
+    // clusterization on this module / surface.
+    dd.threshold().back() = 0.f;
+
+    // Fill the new element with the digitization configuration for the
+    // surface.
+    const auto& binning_data = data.segmentation.binningData();
+    dd.reference_x().back() = binning_data.at(0).min;
+    dd.reference_y().back() = binning_data.at(1).min;
+    dd.pitch_x().back() = binning_data.at(0).step;
+    dd.pitch_y().back() = binning_data.at(1).step;
+    dd.dimensions().back() = data.dimensions;
+    dd.variance_y().back() = data.variance_y;
+}
+
+void read_csv_dd(traccc::silicon_detector_description::host& dd,
+                 std::string_view geometry_file,
+                 const traccc::digitization_config& digi) {
+
+    // Read the geometry description as a map of surface tranformations.
+    const std::map<traccc::geometry_id, traccc::transform3> surfaces =
+        traccc::io::csv::read_surfaces(traccc::io::data_directory() +
+                                       geometry_file.data());
+
+    // Fill the detector description with information about the (sensitive)
+    // surfaces, and the digitization configurations belonging to those
+    // surfaces.
+    dd.reserve(surfaces.size());
+    for (const auto& [geom_id, transform] : surfaces) {
+
+        // Acts geometry identifier for the surface.
+        const Acts::GeometryIdentifier acts_geom_id{geom_id};
+
+        // Skip non-sensitive surfaces. They are not needed in the
+        // "detector description".
+        if (acts_geom_id.sensitive() == 0) {
+            continue;
+        }
+
+        // Add a new element to the detector description.
+        dd.resize(dd.size() + 1);
+
+        // Fill the new element with the geometry ID and the transformation of
+        // the surface in question.
+        dd.surface_link().back() = detray::geometry::barcode{geom_id};
```

Comment:
:thinking: I guess for the CSV input I should rather just set some invalid `bardcode` objects. How should I best do this?

Since with a CSV geometry input this variable will not actually be used for anything useful anyway... (As much as I can tell...)

Reply:
It's not the reconstruction code that would care about this variable as it turns out, but some of @beomki-yeo's truth-matching code. If I don't set `surface_link` like this for CSV geometries, one of the unit tests breaks.

```
...
[ RUN      ] mappper.measurement_particle_map_with_clusterization
/data/ssd-1tb/projects/traccc/traccc/tests/io/test_mapper.cpp:288: Failure
Expected equality of these values:
  has_first_cluster
    Which is: false
  true

/data/ssd-1tb/projects/traccc/traccc/tests/io/test_mapper.cpp:290: Failure
Expected equality of these values:
  has_first_particle
    Which is: false
  true

/data/ssd-1tb/projects/traccc/traccc/tests/io/test_mapper.cpp:291: Failure
Expected equality of these values:
  has_second_particle
    Which is: false
  true

/data/ssd-1tb/projects/traccc/traccc/tests/io/test_mapper.cpp:292: Failure
Expected equality of these values:
  has_third_particle
    Which is: false
  true

[  FAILED  ] mappper.measurement_particle_map_with_clusterization (96 ms)
...
```

Since I want to leave the tests as untouched as possible in this PR, this slightly nonsensical setting should remain for now. (As it's consistent with how `traccc::cell_module` objects are set up at the moment with the TML geometry.)

Reply:
Curious! I am suspecting that this is relying on some accidental bitwise compatibility of the integer values here? Sounds a bit sketchy if it is that way, but I don't understand the code well enough to know